[22.11] curl: fix noproxy tailmatch issue

tksldk opened the pull request:
###### Description of changes
Fixes https://github.com/curl/curl/issues/9821

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt commented on pkgs/tools/networking/curl/7.85.0-noproxy-tailmatch.patch:
This should probably be `7.87.0-noproxy-tailmatch.patch`, since this the release this fix is taken from. The regression came up in 7.86.0, so I'm not sure where you take 7.85.0 from.

mweinelt commented on pkgs/tools/networking/curl/7.85.0-noproxy-tailmatch.patch:
```diff
@@ -0,0 +1,206 @@
+     { "foobar", "foobar", TRUE},
+     { "192.168.0.1", "foobar", FALSE},
+
+From e0c3496c3842fe2a84297d10a4ec52794d58bd87 Mon Sep 17 00:00:00 2001
+From: Daniel Stenberg <daniel@haxx.se>
+Date: Sun, 6 Nov 2022 23:19:51 +0100
+Subject: [PATCH 3/3] fixup add another line to unit1614
+
+---
+ tests/unit/unit1614.c | 1 +
+ 1 file changed, 1 insertion(+)
+
+diff --git a/tests/unit/unit1614.c b/tests/unit/unit1614.c
+index 5577c9e6f7a64..523d102bf29d9 100644
+--- a/tests/unit/unit1614.c
++++ b/tests/unit/unit1614.c
+@@ -86,6 +86,7 @@ UNITTEST_START
+     { "example.com", "localhost,example.com,.example.de", TRUE},
+     { "example.com.", "localhost,example.com,.example.de", TRUE},
+     { "nexample.com", "localhost,example.com,.example.de", FALSE},
++    { "www.example.com", "localhost,example.com,.example.de", TRUE},
+     { "foobar", "barfoo", FALSE},
+     { "foobar", "foobar", TRUE},
+     { "192.168.0.1", "foobar", FALSE},
```
This commit does not exist in the upstream tree, instead it became part of the previous commit.

https://github.com/curl/curl/commit/b1953c1933b369b1217ef0f16053e26da63488c3

mweinelt commented on pkgs/tools/networking/curl/default.nix:
```diff
@@ -58,6 +58,7 @@ stdenv.mkDerivation (finalAttrs: {
   };
 
   patches = [
+    ./7.85.0-noproxy-tailmatch.patch
```
```suggestion
    ./7.87.0-noproxy-tailmatch.patch
```

mweinelt commented:
The commit message also does not follow the contributing guidelines. It should start with the package attribute.

tksldk committed:
curl: fix noproxy tailmatch issue

Fixes https://github.com/curl/curl/issues/9821

 7.86.0 has problem in handling --noproxy variable: If the host name is an IP address and the noproxy string contained that
IP address with a following comma, it would erroneously not match.
